lightning/src/ln/channelmanager.rs

```diff
@@ -7903,28 +7952,150 @@ where
 				};
 				if let Some(msg_send_event) = msg_send_event_opt {
 					peer_state.pending_msg_events.push(msg_send_event);
+				};
+				if let Some(mut signing_session) = signing_session_opt {
+					match chan_phase_entry.get_mut() {
+						ChannelPhase::UnfundedOutboundV2(chan) => {
+							chan.funding_tx_constructed(&mut signing_session, &self.logger)
+						},
+						ChannelPhase::UnfundedInboundV2(chan) => {
+							chan.funding_tx_constructed(&mut signing_session, &self.logger)
+						},
+						_ => Err(ChannelError::Warn(
+							"Got a tx_complete message with no interactive transaction construction expected or in-progress"
+							.into())),
+					}.and_then(|(commitment_signed, funding_ready_for_sig_event_opt)| {
+							let (channel_id, channel_phase) = chan_phase_entry.remove_entry();
+							match channel_phase {
+								ChannelPhase::UnfundedOutboundV2(chan) => {
+									chan.into_channel(signing_session)
+								},
+								ChannelPhase::UnfundedInboundV2(chan) => {
+									chan.into_channel(signing_session)
+								},
+								_ => {
+									debug_assert!(false); // It cannot be another variant as we are in the `Ok` branch of the above match.
+									Err(ChannelError::Warn(
+										"Got a tx_complete message with no interactive transaction construction expected or in-progress"
+											.into()))
+								}
+							}.map(|channel| (channel_id, channel, funding_ready_for_sig_event_opt, commitment_signed))
+						}).map(|(channel_id, channel, funding_ready_for_sig_event_opt, commitment_signed)| {
+							peer_state.channel_by_id.insert(channel_id, ChannelPhase::Funded(channel));
+							if let Some(funding_ready_for_sig_event) = funding_ready_for_sig_event_opt {
+								let mut pending_events = self.pending_events.lock().unwrap();
+								pending_events.push_back((funding_ready_for_sig_event, None));
+							}
+							peer_state.pending_msg_events.push(events::MessageSendEvent::UpdateHTLCs {
+								node_id: counterparty_node_id,
+								updates: CommitmentUpdate {
+									commitment_signed,
+									update_add_htlcs: vec![],
+									update_fulfill_htlcs: vec![],
+									update_fail_htlcs: vec![],
+									update_fail_malformed_htlcs: vec![],
+									update_fee: None,
+								},
+							});
+						}).map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))
```

Comment:
When possible, it's best to avoid side-effects inside `map` and `and_then` calls. It would make the code easier to follow and less indented. Consider the following:

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index b23bd08ff..c2e67287a 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -7955,7 +7955,7 @@ where
                                };
                                if let Some(mut signing_session) = signing_session_opt {
                                        let funding_txid = signing_session.unsigned_tx.txid();
-                                       match chan_phase_entry.get_mut() {
+                                       let (commitment_signed, funding_ready_for_sig_event_opt) = match chan_phase_entry.get_mut() {
                                                ChannelPhase::UnfundedOutboundV2(chan) => {
                                                        chan.funding_tx_constructed(&mut signing_session, &self.logger)
                                                },
@@ -7965,45 +7965,39 @@ where
                                                _ => Err(ChannelError::Warn(
                                                        "Got a tx_complete message with no interactive transaction construction expected or in-progress"
                                                        .into())),
-                                       }.and_then(|(commitment_signed, funding_ready_for_sig_event_opt)| {
-                                                       let (channel_id, channel_phase) = chan_phase_entry.remove_entry();
-                                                       match channel_phase {
-                                                               ChannelPhase::UnfundedOutboundV2(chan) => {
-                                                                       chan.into_channel(signing_session)
-                                                               },
-                                                               ChannelPhase::UnfundedInboundV2(chan) => {
-                                                                       chan.into_channel(signing_session)
-                                                               },
-                                                               _ => {
-                                                                       debug_assert!(false); // It cannot be another variant as we are in the `Ok` branch of the above match.
-                                                                       Err(ChannelError::Warn(
-                                                                               "Got a tx_complete message with no interactive transaction construction expected or in-progress"
-                                                                                       .into()))
-                                                               }
-                                                       }.map(|channel| (channel_id, channel, funding_ready_for_sig_event_opt, commitment_signed))
-                                               }).map(|(channel_id, mut channel, funding_ready_for_sig_event_opt, commitment_signed)| {
-                                                       channel.set_next_funding_txid(&funding_txid);
-                                                       peer_state.channel_by_id.insert(channel_id, ChannelPhase::Funded(channel));
-                                                       if let Some(funding_ready_for_sig_event) = funding_ready_for_sig_event_opt {
-                                                               let mut pending_events = self.pending_events.lock().unwrap();
-                                                               pending_events.push_back((funding_ready_for_sig_event, None));
-                                                       }
-                                                       peer_state.pending_msg_events.push(events::MessageSendEvent::UpdateHTLCs {
-                                                               node_id: counterparty_node_id,
-                                                               updates: CommitmentUpdate {
-                                                                       commitment_signed,
-                                                                       update_add_htlcs: vec![],
-                                                                       update_fulfill_htlcs: vec![],
-                                                                       update_fail_htlcs: vec![],
-                                                                       update_fail_malformed_htlcs: vec![],
-                                                                       update_fee: None,
-                                                               },
-                                                       });
-                                               }).map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))
-                               } else {
-                                       // We're not in a signing session yet so we don't need to do anything else.
-                                       Ok(())
+                                       }.map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))?;
+
+                                       let (channel_id, channel_phase) = chan_phase_entry.remove_entry();
+                                       let mut channel = match channel_phase {
+                                               ChannelPhase::UnfundedOutboundV2(chan) => chan.into_channel(signing_session),
+                                               ChannelPhase::UnfundedInboundV2(chan) => chan.into_channel(signing_session),
+                                               _ => {
+                                                       debug_assert!(false); // It cannot be another variant as we are in the `Ok` branch of the above match.
+                                                       Err(ChannelError::Warn(
+                                                               "Got a tx_complete message with no interactive transaction construction expected or in-progress"
+                                                                       .into()))
+                                               },
+                                       }.map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))?;
+
+                                       channel.set_next_funding_txid(&funding_txid);
+                                       peer_state.channel_by_id.insert(channel_id, ChannelPhase::Funded(channel));
+                                       if let Some(funding_ready_for_sig_event) = funding_ready_for_sig_event_opt {
+                                               let mut pending_events = self.pending_events.lock().unwrap();
+                                               pending_events.push_back((funding_ready_for_sig_event, None));
+                                       }
+                                       peer_state.pending_msg_events.push(events::MessageSendEvent::UpdateHTLCs {
+                                               node_id: counterparty_node_id,
+                                               updates: CommitmentUpdate {
+                                                       commitment_signed,
+                                                       update_add_htlcs: vec![],
+                                                       update_fulfill_htlcs: vec![],
+                                                       update_fail_htlcs: vec![],
+                                                       update_fail_malformed_htlcs: vec![],
+                                                       update_fee: None,
+                                               },
+                                       });
                                }
+                               Ok(())
                        },
                        hash_map::Entry::Vacant(_) => {
                                Err(MsgHandleErrInternal::send_err_msg_no_close(format!("Got a message for a channel from the wrong node! No such channel for the passed counterparty_node_id {}", counterparty_node_id), msg.channel_id))
```